File: backend/app/schemas/order.py
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class OrderCreate(BaseModel):
    symbol: str
    side: Literal['buy', 'sell']
    order_type: Literal['market', 'limit']
    quantity: int = Field(gt=0)
    price: Decimal = Field(gt=0)
    stop_loss_price: Decimal | None = Field(default=None, gt=0)
    take_profit_price: Decimal | None = Field(default=None, gt=0)


class OrderRead(BaseModel):
    id: int
    tenant_id: str
    account_id: int
    symbol: str
    name: str | None = None
    side: Literal['buy', 'sell']
    order_type: Literal['market', 'limit']
    status: Literal['pending', 'accepted', 'filled', 'rejected', 'cancelled', 'expired']
    quantity: int
    price: Decimal
    filled_quantity: int
    filled_price: Decimal
    reject_reason: str | None
    risk_rule_version: str | None = None
    correlation_id: str | None = None
    strategy_id: int | None = None
    strategy_run_id: int | None = None

    model_config = ConfigDict(from_attributes=True, use_enum_values=True)

    @field_validator("side", "order_type", "status", mode="before")
    @classmethod
    def _coerce_enum_values(cls, value):
        return getattr(value, "value", value)

File: frontend/src/types/order.ts
export interface OrderItem {
  id: number
  tenant_id: string
  account_id: number
  symbol: string
  name: string | null
  side: 'buy' | 'sell'
  order_type: 'market' | 'limit'
  status: 'pending' | 'filled' | 'rejected' | 'cancelled'
  quantity: number
  price: string
  filled_quantity: number
  filled_price: string
  reject_reason: string | null
  risk_rule_version: string | null
}

export interface CreateOrderPayload {
  symbol: string
  side: 'buy' | 'sell'
  order_type: 'market' | 'limit'
  quantity: number
  price: number
  stop_loss_price?: number | null
  take_profit_price?: number | null
}

export interface CreateOrderResponse {
  status: 'accepted' | 'rejected'
  risk_rule_version?: string | null
  rejection_reason?: string
  order?: {
    id: number
    symbol: string
    name?: string | null
    quantity: number
    price: number
    status: string
    reject_reason?: string | null
  }
}

export interface CancelOrderResponse {
  status: 'accepted' | 'rejected' | 'not_found'
  message?: string
}

export interface MatchPendingOrdersResponse {
  status: 'accepted'
  matched_count: number
  matched_orders: Array<{
    id: number
    symbol: string
    name?: string | null
    status: string
    risk_rule_version: string | null
    filled_price: number
    filled_quantity: number
  }>
}

export type OrderStatus = OrderItem['status']
export type OrderSide = OrderItem['side']
export type OrderType = OrderItem['order_type']
